src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsReadStream.cs

```diff
@@ -0,0 +1,340 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers.Binary;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+
+namespace Microsoft.Data.SqlClientX.IO
+{
+    /// <summary>
+    /// A stream which handles reading TDS messages from the wire, and extracting the data from the 
+    /// packets in the message packets.
+    /// </summary>
+    internal class TdsReadStream : Stream, ITdsReadStream
+    {
+        // TODO: Handle Cancellation tokens in all async paths.
+        #region Private Fields
+
+        private Stream _underlyingStream;
+
+        // The buffer to hold the TDS read data.
+        private byte[] _readBuffer;
+
+        // The read pointer inside the buffer.
+        private int _readIndex = 0;
+
+        // The end of the data index in the buffer.
+        private int _readBufferDataEnd = 0;
+
+        /// <summary>
+        /// The number of bytes left in the packet to be consumed.
+        /// This is not necessarily the number of bytes available in the buffer.
+        /// </summary>
+        private int _packetDataLeft = 0;
+
+        /// <summary>
+        /// The number of bytes in the header, according to the header.
+        /// </summary>
+        private int _packetHeaderDataLength = 0;
+
+        private byte _packetHeaderType = 0;
```

Comment:
Wouldn't this need to be exposed via a property? I'm thinking whoever is consuming the packets will want to be able to verify the type of packet that's available.

Reply:
Well, in the current implementation of the driver, this is never read.
However, for the completeness of this Stream, I will expose a property.